test/e2e/framework/pod/exec_util.go

```diff
@@ -139,6 +152,12 @@ func ExecShellInPodWithFullOutput(ctx context.Context, f *framework.Framework, p
 	return execCommandInPodWithFullOutput(ctx, f, podName, "/bin/sh", "-c", cmd)
 }
 
+// ExecShellInContainerWithFullOutput executes the specified command on the Pod and returns stdout, stderr and error.
+func ExecShellInContainerWithFullOutput(ctx context.Context, f *framework.Framework, podName string, containerName string, cmd string) (string, string, error) {
```

Comment:
Why can't you do this with `ExecCommandInContainerWithFullOutput`?

Reply:
why makes it easier to read?

It will be better to avoid exposing new functions on this libraries if not necessary, increase the maintainability cost for small return

Reply:
> why makes it easier to read?

Because it follows existing pattern, why existing pattern makes it easier? Without ExecShell* functions addition of "/bin/sh", "-c" is required to be added as command argument whenever execCommand* is used.

> It will be better to avoid exposing new functions on this libraries if not necessary, increase the maintainability cost for small return

I agree , function `execCommandInContainerWithFullOutput ` was added to execute command to specific container, using NewPodClient interface. When i tried to use ExecCommandInContainerWithFullOutput with make test-e2e-node running against remote machine, in my setup, it failed because it wasn't able to find kubectl. Insipiration of this function came looking in my logs that `execCommandInPod` called by `ExecShellInPod` worked fine.

Reply:
> addition of "/bin/sh", "-c" 

that is not a big problem , is it?

Reply:
Pattern is used in [28 files according to cs](https://cs.k8s.io/?q=ExecShell&i=nope&files=&excludeFiles=&repos=)

Reply:
let me try to understand, so you are saying that you prefer to add a new method instead of using `ExecCommandInContainerWithFullOutput` because there are other 25 files that build a command in that way, and the only difference is that we need to append `"/bin/sh", "-c", cmd` in the beginning ?

Let me give you context, everytime something changes in the framework this e2e forces us to modify a lot of functions in a single PR https://github.com/kubernetes/kubernetes/pull/114783 with a high risk of regressions or subtle bugs, so expanding the existing helpers keeps adding on this problem ... 

this seems a cosmetic benefit and increases the helpers function to maintain,

Reply:
> let me try to understand, so you are saying that you prefer to add a new method instead of using `ExecCommandInContainerWithFullOutput` because there are other 25 files that build a command in that way, and the only difference is that we need to append `"/bin/sh", "-c", cmd` in the beginning ?

No, what i am saying is that `ExecCommandInContainerWithFullOuput` cannot be used because it uses Kubectl. For PodResize tests to pass, i had to introduce `execCommandInContainerWithFullOutput` which performs the same task as `ExecCommandInContainerWithFullOuput` using PodClient instead of kubectl.

In my lab `ExecCommandInContainerWithFullOuput` unfortunately it fails, If there is a mistake from my side how i run the tests please let me know ( i have shared the command i use in the PR description )

> Let me give you context, everytime something changes in the framework this e2e forces us to modify a lot of functions in a single PR #114783 with a high risk of regressions or subtle bugs, so expanding the existing helpers keeps adding on this problem ...

Thanks for your comment, if there is a way avoiding the introduction of a new function and be able to execute the pod-resize-test i am fine with it. Unfortunately i wasn't able to find a better solution the last week working on this Issue.

> this seems a cosmetic benefit and increases the helpers function to maintain,

For the `ExecShellInContainerWithFullOutput` i agree it could be interpreted as cosmetic benefit, i interpret it also as an attempt to use same pattern. It `ExecShellInContainerWithFullOutput` is not crucial for this PR `ExecCommandInContainerWithFullOuput` it is so far.

Reply:
I agree that having a separate ExecShellInContainerWithFullOutput doesn't seem worth it. It's not just the maintenance aspect: for someone looking at e2epod, the cognitive burden is higher when they have to read the description of two functions. Even worse, the description of both functions is *the same* (could be fixed of course, but still...).

Perhaps add a comment to `ExecCommandInContainerWithFullOutput` about using `"/bin/sh", "-c", cmd` when shell interpretation of a command string is desired?

Reply:
Thanks i will push an update replacing ExecShell, just for clarity Patrick,  I just did again a test replacing ExecShellInContainerWithFullOutput with ExecCommandInContainerWithFullOutput, test failed. Replacing with **e**xecCommandInContainerWithFullOutput ( please note the small e, it works fine )

https://github.com/kubernetes/kubernetes/blob/0e47e53bba235e954e420bd6118b3ee6ddac4fec/test/e2e_node/pod_resize_test.go#L370

Reply:
> which performs the same task as ExecCommandInContainerWithFullOuput using PodClient instead of kubectl.

ok, sorry, I got confused with the similarity of names and the different discussions,

Reply:
Updated commit, renaming function to `ExecCommandInContainerWithFullOutputUsingPodClient` and removing `ExecShellInContainerWithFullOutput`. Thanks for the feedback.